src/Avalonia.X11/X11PlatformLifetimeEvents.cs

```diff
@@ -0,0 +1,259 @@
+#nullable enable
+using System;
+using System.Collections.Concurrent;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Avalonia.Controls.ApplicationLifetimes;
+using Avalonia.Logging;
+using Avalonia.Platform;
+using Avalonia.Threading;
+
+namespace Avalonia.X11
+{
+    public unsafe class X11PlatformLifetimeEvents : IDisposable, IPlatformLifetimeEventsImpl
+    {
+        private const ulong SmcSaveYourselfProcMask = 1L;
+        private const ulong SmcDieProcMask = 2L;
+        private const ulong SmcSaveCompleteProcMask = 4L;
+        private const ulong SmcShutdownCancelledProcMask = 8L;
+
+        private static readonly ConcurrentDictionary<IntPtr, X11PlatformLifetimeEvents> s_nativeToManagedMapper =
+            new ConcurrentDictionary<IntPtr, X11PlatformLifetimeEvents>();
+
+        private static readonly SMLib.SmcSaveYourselfProc s_saveYourselfProcDelegate = SmcSaveYourselfHandler;
+        private static readonly SMLib.SmcDieProc s_dieDelegate = SmcDieHandler;
+
+        private static readonly SMLib.SmcShutdownCancelledProc
+            s_shutdownCancelledDelegate = SmcShutdownCancelledHandler;
+
+        private static readonly SMLib.SmcSaveCompleteProc s_saveCompleteDelegate = SmcSaveCompleteHandler;
+        private static readonly SMLib.SmcInteractProc s_smcInteractDelegate = StaticInteractHandler;
+        private static readonly SMLib.SmcErrorHandler s_smcErrorHandlerDelegate = StaticErrorHandler;
+        private static readonly ICELib.IceErrorHandler s_iceErrorHandlerDelegate = StaticErrorHandler;
+        private static readonly ICELib.IceIOErrorHandler s_iceIoErrorHandlerDelegate = StaticIceIOErrorHandler;
+        private static readonly SMLib.IceWatchProc s_iceWatchProcDelegate = IceWatchHandler;
+
+        private bool AllowShutdownCancellation =>
+            Environment.GetEnvironmentVariable("AVALONIA_X11_USE_SESSION_MANAGER") != "0";
```

Comment:
Wouldn't it be better to add a configuration in AppBuilder instead of an environment variable?

Reply:
No, this is only intended to be like other X11 switches we already have. @kekekeks and @danwalmsley made this suggestion.

Reply:
@jmacato that switch should live in `X11PlatformOptions` with _default_ value determined by environment variable.

Reply:
gotcha